src/runtime/task/mod.rs

```diff
@@ -1,45 +1,101 @@
+use crate::runtime::execution::ExecutionState;
+use crate::runtime::thread;
+use crate::runtime::thread::continuation::{ContinuationPool, PooledContinuation};
 use bitvec::prelude::*;
 use bitvec::vec::BitVec;
 use futures::future::BoxFuture;
 use futures::task::Waker;
 use std::cell::RefCell;
 use std::fmt::Debug;
 use std::rc::Rc;
+use std::task::{Context, Poll};
 
 pub(crate) mod waker;
 use waker::make_waker;
 
 // A note on terminology: we have competing notions of threads floating around. Here's the
 // convention for disambiguating them:
 // * A "thread" is a user-level unit of concurrency. User code creates threads, passes data
-//   between them, etc. There is no notion of "thread" inside the Shuttle executor, which only
-//   understands Futures. We implement threads via `ThreadFuture`, which emulates a thread inside
-//   a Future using a continuation.
+//   between them, etc.
 // * A "future" is another user-level unit of concurrency, corresponding directly to Rust's notion
 //   in std::future::Future. A future has a single method `poll` that can be used to resume
-//   executing its computation.
+//   executing its computation. Both futures and threads are implemented in AtomicTask,
+//   which wraps a continuation that is resumed when the task is scheduled.
 // * A "task" is the Shuttle executor's reflection of a user-level unit of concurrency. Each task
-//   has a corresponding Future, which is the user-level code it runs, as well as a state like
+//   has a corresponding continuation, which is the user-level code it runs, as well as a state like
 //   "blocked", "runnable", etc. Scheduling algorithms take as input the state of all tasks
 //   and decide which task should execute next. A context switch is when one task stops executing
 //   and another begins.
 // * A "continuation" is a low-level implementation of green threading for concurrency. Each
-//   ThreadFuture contains a corresponding continuation. When the Shuttle executor polls a
-//   ThreadFuture, which corresponds to a user-level thread, the ThreadFuture resumes its
+//   AtomicTask contains a corresponding continuation. When the Shuttle executor steps an
+//   AtomicTask, which corresponds to a user-level thread, the task resumes its
 //   continuation and runs it until that continuation yields, which happens when its thread decides
 //   it might want to context switch (e.g., because it's blocked on a lock).
 
 pub(crate) const DEFAULT_INLINE_TASKS: usize = 16;
 
+#[derive(Debug)]
+pub(crate) enum TaskResult {
+    Yielded,
+    Finished,
+}
+
+pub(crate) struct AtomicTask {
```

Comment:
So now I'm wondering whether `AtomicTask` needs to exist at all :-)

Could we just inline the continuation and task type into `Task` itself, so `Task.inner` becomes `Rc<RefCell<PooledContinuation>>`, and `Execution::step` just calls `inner.resume()`? What do you think? Do we gain much from this abstraction?

Reply:
I tried doing something like this, but ended up changing a lot of unrelated stuff since I didn't wrap the `PooledContinuation` like that. I think this would be fine. The abstraction lets us construct `AtomicTasks`s in the library, then `Task`s in the execution loop, which doesn't really matter as far as I can tell.

Reply:
+1 to James's comment.  It seems that now `Task` is the only abstraction we need.